### PR TITLE
Redis cluster client

### PR DESCRIPTION
**Pull request by h4lflife**
Redis cluster client.

This is implementated as a wrapper over the existing resty-redis client adding cluster functionality. Basic functionality except pipelining is working. Working on pipelining support.
- Each cluster is identified by a "cluster id". 
- Cluster representation identified by "cluster id" is shared across all requests within a worker process. 
- Supports multiple redis clusters. 
- Connection pooling as present in the existing resty-redis client is available.
- Performance nearly same as original resty-redis under normal conditions.

ref: https://github.com/antirez/redis-rb-cluster

Example:

``` lua
local redis_cluster = require("redis_cluster")

local cluster_id = "test_cluster"

-- Subset of nodes within the cluster
local startup_nodes = { 
    {"127.0.0.1", 7004}, 
    {"127.0.0.1", 7000}, 
    {"127.0.0.1", 7001}
}

local opt = { 
    timeout = 100,
    keepalive_size = 100,
    keepalive_duration = 60000
}

local rc = redis_cluster:new(cluster_id, startup_nodes, opt)

rc:initialize()

local ok, err = rc:set("key1", "val1")
if not ok then
    ngx.say("Unable to set key1: ", err)
else
    ngx.say("key1 set result: ", ok) 
end

local res, err = rc:get("key1")
if not res then
    ngx.say("Failed to get key1: ", err)
else
    ngx.say("key1:", res)
end

-- (same as above, slightly faster)
res, err = rc:send_cluster_command("get", "key1")
if not res then
    ngx.say("Failed to get key1 with send_cluster_command: ", err)
else
    ngx.say("key1 using send_cluster_command:", res)
end
```

Open Issues:
- At the time of initialization(once per worker) and cluster reconfiguration, all requests during the time window of the data refresh(should be quite small, few ms) will try to update the cluster representation. It doesn't seem to have an effect on functionality but the condition needs to be removed as it would cause a small spike in latency during cluster reconfiguration. Would like to have opinion on this one.
- Pipeline "asking" request during cluster reconfiguration.

Please review and let me know of any comments.

### CODE REVIEW

**Comment by misiek08** [comment]
It's old pull request but I have question, because I don't know if I understand your code correctly.

You have hard-coded limit for max. 20 clusters and 500 nodes in each cluster, right?
If yes, I think one call for your code should be one cluster, and the nodes should be stored in dynamically sized array.

**Comment by zhduan** [comment]
@agentzh @h4lflife  Are there any current activities on this pull request?  We need the cluster support too and love to see it moved forward.

**Comment by agentzh** [comment]
@zhduan My hunch is that it's better to be implemented in a separate wrapper library, in the same spirit of @pintsized's [lua-resty-redis-connector](https://github.com/pintsized/lua-resty-redis-connector)

File: lib/resty/redis_cluster.lua [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+local cluster_invalid_cmds = {
+    ["info"] = true,
+    ["multi"] = true,
+    ["exec"] = true,
+    ["slaveof"] = true,
+    ["config"] = true,
+    ["shutdown"] = true
+}
+
+local band, bor, bxor = bit.band, bit.bor, bit.bxor
```
Comment (wangmengcn) [inline lib/resty/redis_cluster.lua]: hi,i'm a newer of github,when i use your redis_cluster, i met a problem at this line:attempt to index local 'bit' (a boolean value), i'm wondering how to solve this problem,can you help me? thanks a lot!

Comment (wangmengcn) [inline lib/resty/redis_cluster.lua]: i have solved my problem. it's my own mistake.Now i'm using redis_cluster with redis3,0.1 and it works pretty well.

**Comment by fankeke** [comment]
94 lua entry thread aborted: runtime error: ../lua_file/cluster_redis/lib/resty/redis_cluster.lua:170: attempt to call method 'cluster' (a nil value)

when called r:cluster(xxx) (r is redis instance in redis.lua) .

anything wrong?

**Comment by wyTrivail** [comment]
thanks for implementing it, but i have a problem that whether there are connection pool in your code?

File: lib/resty/redis_cluster.lua [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                        local host_port = string_split(addr_str, ":", 2)
+                        addr = { host_port[1], 
+                                 tonumber(host_port[2]), 
+                                 addr_str }
+                    end
+                    cluster.nodes[#(cluster.nodes) + 1] = addr
+                    
+                    local cluster_slots = cluster.slots
+
+                    for slot_index = 9, #fields do
```
Comment (wyTrivail) [inline lib/resty/redis_cluster.lua]: here，if the slots in one node is 0-1665 5461-7127 10923-12588,  there will be some lose slots when add slot cache

**Comment by jzh800** [comment]
Whether to support password access?
like the  red:auth("foobared").

**Comment by wjs57y** [comment]
请问怎么配置验证密码

**Comment by agentzh** [comment]
@wjs57y Please, no Chinese here. This place is considered English only. It is especially rude to reply to a pure English issue thread. If you really
want to use Chinese, please join the openresty (Chinese) mailing list instead.
Please see https://openresty.org/en/community.html Thanks for your cooperation.

**Comment by harryin777** [comment]
Here is a error msg : Uninitialized cluster. I follow the Example, what`s wrong? Is there anyone have a same problem?

**Pull request closed without merging**